File: java21/src/Nif.java
import java.io.File;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

public class Nif {

    public static List<Vertex> parse(File file) throws IOException {
        var buf = ByteBuffer.wrap(Files.readAllBytes(file.toPath()))
                            .order(ByteOrder.LITTLE_ENDIAN);

        buf.position(48); // skip NIF file header

        var numBlocks     = buf.getInt();
        buf.position(buf.position() + 10);
        var numBlockTypes = Short.toUnsignedInt(buf.getShort());
        buf.getInt(); // unknown

        for (int i = 0; i < numBlockTypes; i++) {
            readString(buf);
            skipZeros(buf);
        }

        for (int i = 0; i < numBlocks; i++) buf.getShort(); // block type indices

        for (int i = 0; i < 3; i++) skipToRef(buf); // skip to NiTriStripsData
        skipZeros(buf);

        var numVertices = Short.toUnsignedInt(buf.getShort());
        buf.getShort(); // unknown
        buf.get();      // has_vertices

        var vertices = new Vertex[numVertices];
        for (int i = 0; i < numVertices; i++)
            vertices[i] = new Vertex(buf.getFloat(), buf.getFloat(), buf.getFloat());

        skipToRef(buf);

        buf.getShort(); // num_triangles
        buf.getShort(); // num_strips
        var stripSize = Short.toUnsignedInt(buf.getShort());
        buf.get();      // unknown byte

        var result = new ArrayList<Vertex>(stripSize);
        for (int i = 0; i < stripSize; i++) {
            var idx = Short.toUnsignedInt(buf.getShort());
            if (idx < vertices.length) result.add(vertices[idx]);
        }

        return result;
    }

    private static String readString(ByteBuffer buf) {
        var sb = new StringBuilder();
        while (buf.hasRemaining() && buf.get(buf.position()) > 32)
            sb.append((char) buf.get());
        if (buf.hasRemaining()) buf.get(); // consume terminator
        return sb.toString();
    }

    private static void skipZeros(ByteBuffer buf) {
        while (buf.hasRemaining() && buf.get(buf.position()) == 0) buf.get();
    }

    private static void skipToRef(ByteBuffer buf) {
        while (buf.remaining() >= 4) {
            if (buf.get(buf.position())   == (byte) -1 && buf.get(buf.position()+1) == (byte) -1 &&
                buf.get(buf.position()+2) == (byte) -1 && buf.get(buf.position()+3) == (byte) -1) {
                buf.position(buf.position() + 4);
                return;
            }
            buf.get();
        }
    }
}

File: kotlin/src/main/kotlin/Nif.kt
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder

data class Vertex(val x: Float, val y: Float, val z: Float)

fun parseNif(file: File): List<Vertex> {
    val buf = ByteBuffer.wrap(file.readBytes()).order(ByteOrder.LITTLE_ENDIAN)

    buf.position(48) // skip NIF file header

    val numBlocks     = buf.int
    buf.position(buf.position() + 10)
    val numBlockTypes = buf.short.toUShort().toInt()
    buf.int // unknown

    repeat(numBlockTypes) {
        buf.readString()
        buf.skipZeros()
    }

    repeat(numBlocks) { buf.short } // block type index per block

    repeat(3) { buf.skipToRef() }  // skip to NiTriStripsData
    buf.skipZeros()

    val numVertices = buf.short.toUShort().toInt()
    buf.short  // unknown
    buf.get()  // has_vertices

    val vertices = Array(numVertices) { Vertex(buf.float, buf.float, buf.float) }

    buf.skipToRef()

    buf.short  // num_triangles
    buf.short  // num_strips
    val stripSize = buf.short.toUShort().toInt()
    buf.get()  // unknown byte

    return (0 until stripSize).mapNotNull { vertices.getOrNull(buf.short.toUShort().toInt()) }
}

private fun ByteBuffer.readString(): String {
    val sb = StringBuilder()
    while (hasRemaining() && get(position()) > 32) sb.append(get().toInt().toChar())
    if (hasRemaining()) get() // consume terminator
    return sb.toString()
}

private fun ByteBuffer.skipZeros() {
    while (hasRemaining() && get(position()) == 0.toByte()) get()
}

private fun ByteBuffer.skipToRef() {
    val ref = byteArrayOf(-1, -1, -1, -1)
    while (remaining() >= 4) {
        if (get(position()) == ref[0] && get(position()+1) == ref[1] &&
            get(position()+2) == ref[2] && get(position()+3) == ref[3]) {
            position(position() + 4)
            return
        }
        get()
    }
}
